### PR TITLE
Move suppressions to `core`

### PR DESCRIPTION
**Pull request by BraisGabin**
All the suppression logic should be handled by the `core`. Rule shouldn't know about that. This PR moves that logic to `core` and also moves the functions that are no longer used in `api` to `core` to keep `api` as clean as possible.

### CODE REVIEW

File: detekt-test/src/main/kotlin/io/gitlab/arturbosch/detekt/test/RuleExtensions.kt [hunk 1]
```diff
@@ -25,12 +27,12 @@ fun Rule.compileAndLint(@Language("kotlin") content: String): List<Finding> {
 
 fun Rule.lint(@Language("kotlin") content: String): List<Finding> {
     val ktFile = compileContentForTest(content)
-    return visitFile(ktFile)
+    return visitFile(ktFile).filterSuppressed(this)
```
Comment (BraisGabin) [inline detekt-test/src/main/kotlin/io/gitlab/arturbosch/detekt/test/RuleExtensions.kt:30]: I don't like to add this here too much but I think that for now it's the simpler solution.

Why do we need this? Right now we have some tests on the rules that check that when an annotation is set the finding is not raised. Those tests are testing two things. From one side they are testing that the suppression works, that's tested already somewhere else so it should not be tested. But they are also testing that the `KtElement` that they provide is correct or that the configured `alias` is correct. We could refactor those tests but the tests would be more difficult to follow (I tried). For that reason I decide to keep this here.

The good part is that, if we want, we could refactor those tests and remove this later.

**Comment by BraisGabin** [comment]
@detekt/maintainers could I get a review here?

File: detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt [hunk 2]
```diff
@@ -101,15 +98,9 @@ open class Rule(
 
     /**
      * Reports a single code smell finding.
```
Comment (schalkms) [inline detekt-api/src/main/kotlin/io/gitlab/arturbosch/detekt/api/Rule.kt:100]: At this point, detekt does not report, bur rather add it to the findings list.

File: detekt-test/src/main/kotlin/io/gitlab/arturbosch/detekt/test/RuleExtensions.kt (hunk 1 above)
Comment (schalkms) [reply detekt-test/src/main/kotlin/io/gitlab/arturbosch/detekt/test/RuleExtensions.kt:30]: Step-by.step is the way to go here. As you said, I don't think it's ideal approach, though sufficient for now.

File: detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/AnalyzerSpec.kt [hunk 3]
```diff
@@ -384,6 +384,29 @@ class AnalyzerSpec(val env: KotlinCoreEnvironment) {
             }
             assertThat(findings).isEmpty()
         }
+
+        @ParameterizedTest
+        @ValueSource(strings = ["MaxLineLength", "detekt.MaxLineLength", "MLL", "custom", "all"])
+        fun `when the ktElement is suppressed the issue is not raised`(suppress: String) {
```
Comment (schalkms) [inline detekt-core/src/test/kotlin/io/gitlab/arturbosch/detekt/core/AnalyzerSpec.kt:390]: Tests are way cleaner now.

**Review by schalkms — APPROVED** [review]
👍
Good to go in my point of view.
Just the function comment needs to be adapted.

**Pull request merged**